mmengine/_strategy/base.py

```diff
@@ -0,0 +1,924 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import copy
+import os.path as osp
+import platform
+import time
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from typing import Callable, Dict, List, Optional, Sequence, Tuple, Union
+
+import torch
+import torch.nn as nn
+from torch.optim import Optimizer
+
+import mmengine
+from mmengine.config import Config, ConfigDict
+from mmengine.dist import (broadcast, get_dist_info, infer_launcher,
+                           is_distributed)
+from mmengine.logging import MMLogger
+from mmengine.model import revert_sync_batchnorm
+from mmengine.model.wrappers import is_model_wrapper
+from mmengine.optim import (BaseOptimWrapper, OptimWrapperDict,
+                            _ParamScheduler, build_optim_wrapper)
+from mmengine.registry import MODELS, OPTIM_WRAPPERS, PARAM_SCHEDULERS
+from mmengine.utils import digit_version
+from mmengine.utils.dl_utils import (TORCH_VERSION, collect_env,
+                                     set_multi_processing)
+
+ParamSchedulerType = Union[List[_ParamScheduler], Dict[str,
+                                                       List[_ParamScheduler]]]
+
+
+class BaseStrategy(metaclass=ABCMeta):
+    """Base class for all strategies.
+
+    In the process of supporting FSDP, DeepSpeed, and ColossalAI, the
+    scalability of the Runner faced challenges, which led to the redefinition
+    of the Runner's responsibilities. The Strategy abstraction was split out,
+    which is responsible for constructing, initializing, and saving/loading
+    the state of training components such as models, optimizers, and parameter
+    schedulers.
+
+    Warning:
+        This is an experimental feature, and its interface is subject to
+        change.
+
+    Keyword Args:
+        work_dir (str): The working directory to save checkpoints. The logs
+            will be saved in the subdirectory of `work_dir` named
+            :attr:`timestamp`. Defaults to 'work_dirs'.
+        experiment_name (str, optional): Name of current experiment. If not
+            specified, timestamp will be used as :attr:`experiment_name`.
+            Defaults to None.
+        env_kwargs (dict, optional): Environment config passed in
+            :meth:`setup_env`. Defaults to None.
+        log_kwargs (dict, optional): Logger config passed in
+            :meth:`build_logger`. Defaults to None.
```

Comment:
It could be better to provide an example to describe how to configure `log_kwargs` and `env_kwargs`